lighting/pbr.glsl

```diff
@@ -45,65 +46,54 @@ license:
-
-    // Cached
-    Material M  = _mat;
-    if (M.V.x == 0.0 && M.V.y == 0.0 && M.V.z == 0.0) {
-        M.V         = normalize(CAMERA_POSITION - M.position);  // View
-    }
-    M.NoV       = dot(M.normal, M.V);                       // Normal . View
-    M.R         = reflection(M.V, M.normal, M.roughness);   // Reflection
-
-    // TODO: 
-    //  - ScreenSpace Ambient Occlusion when M.ambientOcclusion is not defined
-// #if defined(FNC_SSAO) && defined(SCENE_DEPTH) && defined(RESOLUTION) && defined(CAMERA_NEAR_CLIP) && defined(CAMERA_FAR_CLIP)
-//     vec2 pixel = 1.0/RESOLUTION;
-//     ao = ssao(SCENE_DEPTH, gl_FragCoord.xy*pixel, pixel, 1.);
-// #endif 
-
-    // Global Ilumination ( Image Based Lighting )
-    // ------------------------
-    vec3 E = envBRDFApprox(specularColor, M);
-    float diffuseAO = M.ambientOcclusion;
+vec4 pbr(const Material mat, ShadingData shadingData) {
+    // Shading Data
+    // ------------
+    shadingData.N = mat.normal;
+    shadingData.R = reflection(shadingData.V,  shadingData.N, mat.roughness);
+    shadingData.fresnel = max(mat.f0.r, max(mat.f0.g, mat.f0.b));
+    shadingData.roughness = mat.roughness;
+    shadingData.linearRoughness = mat.roughness;
+    shadingData.diffuseColor = mat.albedo.rgb * (vec3(1.0) - mat.f0) * (1.0 - mat.metallic);
+    shadingData.specularColor = mix(mat.f0, mat.albedo.rgb, mat.metallic);
+    shadingData.NoV = dot(shadingData.N, shadingData.V);
+
+    // Indirect Lights ( Image Based Lighting )
+    // ----------------------------------------
+    vec3 E = envBRDFApprox(shadingData);
+    float diffuseAO = mat.ambientOcclusion;
 
     vec3 Fr = vec3(0.0, 0.0, 0.0);
-    Fr  = envMap(M) * E;
+    Fr  = envMap(mat, shadingData) * E;
     #if !defined(PLATFORM_RPI)
-    Fr  += fresnelReflection(M);
+    Fr  += fresnelReflection(mat, shadingData);
     #endif
-    Fr  *= specularAO(M, diffuseAO);
+    Fr  *= specularAO(mat, shadingData, diffuseAO);
 
-    vec3 Fd = diffuseColor;
+    vec3 Fd = shadingData.diffuseColor;
     #if defined(SCENE_SH_ARRAY)
-    Fd  *= tonemap( sphericalHarmonics(M.normal) );
+    Fd  *= tonemap( sphericalHarmonics(shadingData.N) );
     #else
-    Fd *= envMap(M.normal, 1.0);
+    Fd *= envMap(shadingData.N, 1.0);
     #endif
     Fd  *= diffuseAO;
     Fd  *= (1.0 - E);
 
-    // Local Ilumination
-    // ------------------------
-    vec3 lightDiffuse = vec3(0.0, 0.0, 0.0);
-    vec3 lightSpecular = vec3(0.0, 0.0, 0.0);
-    
+    // Direct Lights
+    // -------------
+
     {
         #if defined(LIGHT_DIRECTION)
         LightDirectional L = LightDirectionalNew();
-        lightResolve(diffuseColor, specularColor, M, L, lightDiffuse, lightSpecular);
         #elif defined(LIGHT_POSITION)
         LightPoint L = LightPointNew();
-        lightResolve(diffuseColor, specularColor, M, L, lightDiffuse, lightSpecular);
         #endif
+        lightResolve(L, mat, shadingData);
```

Comment:
This fails if `LIGHT_DIRECTION` and  `LIGHT_POSITION` are not defined

suggestion
```glsl
        #if defined(LIGHT_DIRECTION)
        LightDirectional L = LightDirectionalNew();
        lightResolve(L, mat, shadingData);

        #elif defined(LIGHT_POSITION)
        LightPoint L = LightPointNew();
        lightResolve(L, mat, shadingData);
        #endif
```

Reply:
Absolutely. Thanks.